doc/smart-switch/PA-Validation/SmartSwitchPAValidationOffload.md

```diff
@@ -0,0 +1,149 @@
+## SmartSwitch PA Validation NPU Offload
+
+- [About this Manual](#about-this-manual)
+- [Definitions/Abbrevations](#definitionsabbrevations)
+- [1. Requirements Overview](#1-requirements-overview)
+  - [1.1 Functional Requirements](#11-functional-requirements)
+  - [1.2 Scale Requirements](#12-scale-requirements)
+- [2 Modules Design](#2-modules-design)
+  - [2.1 Dash Offload Manager](#21-dash-offload-manager)
+    - [2.1.1 SONiC GNMI server changes](#211-sonic-gnmi-server-changes)
+  - [2.2 PA Validation Offload](#22-pa-validation-offload)
+    - [2.2.1 PA Validation Offload ACL configuration](#221-pa-validation-offload-acl-configuration)
+  - [2.3 DPU Shut/Restart](#23-dpu-shutrestart)
+- [3. Test Plan](#3-test-plan)
+
+###### Revision
+
+| Rev |     Date    |       Author          | Change Description                |
+|:---:|:-----------:|:---------------------:|-----------------------------------|
+| 0.1 |  06/10/2024 |     Kumaresh Perumal  | Initial version                  |
+| 0.2 |  08/21/2024 |          Yakiv Huryk  | Added DashOffloadManager         |
+
+
+# About this Manual
+This document provides general information about offloading PA validation feature to NPU in Smartswitch.
+
+# Definitions/Abbrevations
+
+|                          |                                          |
+|--------------------------|------------------------------------------|
+| ACL                      | Access Control List                      |
+| NPU                      | Network Processing Unit                  |
+| DPU                      | Data Processing Unit                     |
+ 
+
+# 1. Requirements Overview
+
+## 1.1 Functional Requirements
+- Provide a generic infrastructure for offloading DASH functionality to the NPU (DashOffloadManager).
+- DashOffloadManager should listen to the DASH configuration and either intercept it for offloading, or forward it to the DPU.
+- Provide PA validation support in NPU when this feature is not supported in DPU pipeline model.
+- Use existing ACL tables and ACL rules design in NPU.
+
+## 1.2 Scale Requirements
+
+# 2 Modules Design
+
+## 2.1 Dash Offload Manager
+The new orchagent application DashOffloadManager will be responsible for DASH offloading logic. It will collect all the needed information for offloading and perform all the relevant configurations.
+To get the DASH configuration that should be offloaded, the DashOffloadManager will act as a transparent ZMQ proxy between the GNMI server and the DPU swss, forwarding all the configuration and intercepting the tables that should be offloaded.

```

Comment:
Why is the ZMQ proxy necessary? Could you subscribe to DPU_APPL_DB instead?

Reply:
The idea is to have control over what is sent to the DPU, which is a more flexible tool for implementing different kinds of offload-related functionality.

In case of PA validation offload, we don't forward the PA Validation table entries to the DPU (since DPU has no use of it). This saves some NPU<->DPU bandwidth and allows to keep the DPU side simple (no entries - nothing to create). 

In the future, we can extend this infrastructure also to alter the content of the configuration (if such a need arises for some other offload feature)

Reply:
GNMI client<->GMMI server<->ZMQ<->DPU orchagent<->NPU APPL_STATE_DB

The DPU orchagent should update the results to the DPU_APPL_STATE_DB on the NPU. If we don't forward the PA Validation table entries to the DPU, the client won't know if this configuration failed or was intercepted.

Reply:
Are you referring to the flow described here https://github.com/sonic-net/SONiC/pull/1759 ?
As I understand, the feedback design is not finalized yet. Once it is, I'll have to add the same logic to the offload flow to also fill the result into the DPU_APPL_STATE_DB for the offloaded (intercepted) PA validation entries. Is it ok?

Reply:
Could you please include this in the design?

Reply:
I am having the same concern here.... This seem to be complicated to implement, because of a few things:

1. The DASH offload manager  or the NPU side DASH orchs need to provide the exact same feedback loop as the swss. If anything changed in swss, they need to be changed as well, which can be easily missed and causing problem.
2. In order to support independent DPU upgrades, each DPU will needs to have its own DASH offload manager and all the DASH orchs in the NPU side.
3. Dependency and object handling can also be a problem. This solution is trying to provide a generic way to handle all the DASH object offloading in the future, but I feel it doesn't really do the job. Explicit PA validation rule might be the simpliest case, where the only thing that we needs to do is to redirect the rules into the NPU side. However, other DASH objects can have dependencies, e.g., Implicit PA validation rules are coming from VNET + CA-PA mappings. In this case, we cannot simply redirect the rules into the NPU side, but have to copy it, because they are also used in the outbound pipeline. 

Overall, I feel Gang is correct. The other way that Gang proposed there is actually much more cleaner and maintainable. All we need is just a if case in the swss, and every other things can be reused, such as feedback loops.

Reply:
@ganglyu , please let us know your thought.

Reply:
I'm not sure if this design is necessary. If the DPU doesn't need the PA validation table and only the NPU requires it, we can configure this table directly for the NPU.